### DISCUSSION TITLE
Eos and htcondor submission

### DISCUSSION THREAD

**Discussion by oponcet** [columnflow/columnflow#538] [Q&A]
Dear ColumnFlow experts,

I’m encountering an issue while using HTCondor submission on lxplus. When I submit my job, I receive the following error:

```
ERROR: EOS Submission is not currently supported by the HTCondor Service:
```

Due to size limitations on /afs, I’ve stored CF_DATA on /eos, while CF_SOFTWARE_BASE and CF_JOB_BASE are located on /afs. Unfortunately, moving CF_DATA to /afs isn’t feasible because of these limitations. Is there a way to configure HTCondor to work with CF_DATA stored on /eos?

I also contacted the IT to ask more storage to be able to save CF_DATA output on /afs and then move it to /eos but it's not possible.

Thanks,

Océane, @gsaha009

**Comment by pkausw** [comment]
Hi @oponcet ,

in principle, this error is a limitation of the CERN IT, see the [docs](https://batchdocs.web.cern.ch/troubleshooting/eos.html#no-eos-submission-allowed). I had a short discussion with @riga , and we think it might be the custom logging of law that points to eos. The responsible code for this in columnflow is [here](https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/framework/remote.py#L612). Right now, these custom logs are saved where the final output is saved (i.e. `CF_DATA`).

You should be able to patch the `htcondor_output_directory` function in your `columnflow_patches.py`. This snippet should do the trick

```python
from columnflow.tasks.framework.remote import HTCondorWorkflow

def patched_htcondor_output_directory(self):
    # either
    return self.local_target(dir=True, store="/path/to/location")
    # or
    return self.local_target(dir=True, fs="name_of_local_fs")

HTCondorWorkflow.htcondor_output_directory = patched_htcondor_output_directory
```

Can you try this in your setup?

Cheers,
Philip

**Reply by oponcet** [reply]
Hi,

Thanks for the answer, If I understood well "/path/to/location" should be in /afs workspace? Because I have the following
error:
```
18:28:52: all: 1, pending: 0 (+0), running: 0 (+0), finished: 0 (+0), retry: 1 (+1), failed: 0 (+0), run3_2022_preEE_nano_cp_tau_v12_limited, data_tau_C
1 failed job(s) in task cf.CalibrateEvents_httcp_config_ana__1__1442e19a3f:
    job: 1, branches: [0], id: dummy_job_id, status: retry, code: None, error: unknown job id, log: /afs/cern.ch/user/o/oponcet/private/analysis/CPinHToTauTau/htcondor_output/analysis_httcp/cf.CalibrateEvents/run3_2022_preEE_nano_cp_tau_v12_limited/data_tau_C/nominal/calib__main/vers0/stdall_0To1.txt
job_id:  dummy_job_id
going to submit 1 htcondor job(s), run3_2022_preEE_nano_cp_tau_v12_limited, data_tau_C
submitted 1/1 job(s)
```
Cheers,

Oceane

**Comment by gsaha009** [comment]
Hello Experts,

I would like to add a comment to this discussion. Being an old cern user, I have some advantage from lxplus. I have 100GB in my afs workspace and that is why I somehow managed to use HTCondor directly from afs. 
Now, the issue is, even if I run on the smallest era 2022PreEE up to plotting task, it occupies almost 40GB of storage. So, basically, I have to transfer the old outputs to eos and launch a new set of jobs.
Regarding the above, I have two questions:

 - Is there any way to make the outputs less heavy? If I delete say Calibration output, probably I can not do further checks on that same version, because it will be looking for the missing outputs.
 - As @oponcet mentioned, I also tried to bypass this problem by using eos, but it did not work. It seems a quite expensive software, and if I am not wrong, lxplus wont be able to handle it. The only workaround left is the eos. 

So, it would be really helpful if you guys have a closer look at this, or do some checks on your own lxplus and eos. We wont be able to develop our analysis framework further if the job submission does not get smooth. In addition, if you guys think that a zoom discussion can solve this issue quickly, please let us know.
Thanks.

Gourab

**Comment by gsaha009** [comment]
Hello @pkausw and others

This is working now, and this is a big relief indeed :) . 
As you all know that our target was to use `eos` to save the large output files e.g. `parquet` files, and in order to do that, when we were doing `source setup.sh <name>` for the first time, we explicitly mentioned that we wanted `data` or `output` dir in `eos`. The installation worked, but condor did not. Perhaps, `HTCondor` could not fetch files from `eos`.
But, now I am keeping the `data` dir in `afs` at the time of installation (all are at their default values), and then explicitly mentioned `eos` in `wlcg_fs` for individual tasks like [this](https://github.com/gsaha009/CPinHToTauTau/blob/dev/law.cfg#L91-L104).
There is one small question though. Following the `law` config from `hhbbtautau` analysis, I was using [this](https://github.com/gsaha009/CPinHToTauTau/blob/dev/law.cfg#L141), but it could not understand `$CF_CERN_USER_FIRSTCHAR` and [wlcg_fs_cernbox](https://github.com/gsaha009/CPinHToTauTau/blob/dev/law.cfg#L149) as well.

Btw, I can save the large files in `eos`, but some minor tuning would be nice.
Thanks :)

**Reply by gsaha009** [reply]
I would like to add one more thing.
In the `law.cfg`, does one need to mention
```py
cf.ProduceColumnsWrapper: wlcg
```
like
```py
cf.ProduceColumns: wlcg
```
?

**Reply by gsaha009** [reply]
An additional query:
I am using CERN `HTCondor` for batch jobs. But, I noticed that in the beginning or in the middle maybe, the workflow switches back to `local`. Is it something related to condor, or there is any smart way to handle this?
Thanks

**Reply by pkausw** [reply]
Hi @gsaha009 and @oponcet ,

sorry for the late reply (vacation time intervened). Glad to hear that the HTCondor workflow is now working for you! It's possible that htcondor jobs on lxplus cannot directly write to eos like it's a local mount. By specifying the output file system as wlcg, you explicitely use gfal with the given protocol (in this case `root://`) to perform the transfer. I think this is an important information for other lxplus users :+1:

Regarding your other questions:
- the `CF_CERN_USER_FIRSTCHAR` flag is an environment variable that is derived from `CF_CERN_USER` and is defined in each of your setups (have a look at the corresponding shell script in the `.setups` directory). It's needed because file systems at CERN are order in terms of the first letter of the user name (in your case `g/gsaha`). So this mechanism enables an entry in the law config that works for multiple users and thus adds a little more flexibility
- Wrapper tasks generally don't have outputs themselves - they simply call other tasks. Consequently, you should define the outputs for the underlying tasks, e.g. `cf.ProduceColumns`
- I'm not sure what you mean with "the workflow switches back to `local`" at some point during the run. There are some tasks that run locally by default, e.g. the merging tasks. You can overwrite this behavior at run time with the corresponding `workflow` option (e.g. use `--cf.MergeReducedEvents-workflow htcondor` for `cf.MergeReducedEvents`) 

Let us know if there are additional open questions - and if not, feel free to close this discussion ;)

**Reply by gsaha009** [reply]
Thanks for the reply
regarding my other questions:
 - but my point is that, if I do not hard-code `CF_CERN_USER_FIRSTCHAR` in the `law.cfg`, it does not work. It can not fetch the info on it's own.
  - Got it
  - Let's say I have submitted jobs to HTCondor for `cf.PlotVariables1D` task. Now, I have noticed that most of the time `calibrateEvents` task starts locally, and then from `selection`, it switches to `condor` ...

**Reply by pkausw** [reply]
Okay! Here some follow-up remarks
- Hm, that is strange. As of commit 2f18bc9672956507ef6cddc7503899bdd95427b7, this should work. Do you have this commit in your local setup?
- Regarding the switching between condor/local: Iirc, there might be an issue that the `workflow` option is propagated throughout the task graph (I seem to remember a discussion with @mafrahm about this, maybe he can comment). Regardless, overwriting the `workflow` argument as I mentioned above should always work. This of course makes the command itself a little longer, but that's not a big issue imho and indeed makes the command more verbose. Did you try setting the option directly for a couple of tasks?

**Comment by pkausw** [comment]
Hi @oponcet @gsaha009 ,

is there any follow up to this discussion? If not, feel free to close it ;)

Cheers,
Philip

**Comment by oponcet** [comment]
Hi @pkausw, thanks for the answer, we can close this discussion.
